### ISSUE TITLE
qbittorrent doesn't see python

### ISSUE THREAD

**Issue by NeoTheFox** [NixOS/nixpkgs#104221]
**Describe the bug**
For the search functionality qBittorrent requires Python. The version from nixpkgs can't find it.

**To Reproduce**
1. Open qBittorrent
2. Press "View" and try to enable "Search Engine"
3. An error message about not having Python will be displayed.

**Expected behavior**
A new tab named "Search" appears in the UI

**Additional context**
Looks like the path to python doesn't get set right during the build

@Anton-Latukha 

**Metadata**
 - system: `"x86_64-linux"`
 - host os: `Linux 5.9.8-xanmod1, NixOS, 20.09.1870.14cf906f33c (Nightingale)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.7`
 - channels(root): `"nixos-20.09.1870.14cf906f33c"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute: qbittorrent
# a list of nixos modules affected by the problem
module:
```

**Comment by Anton-Latukha** [comment]
It was discussed previously, that Python is not needed during the build. It is needed only at runtime - and so it would be requested and needs to be supplied only during runtime.

Inclusion of Python into the closure would achieve nothing, since qBittorrent would search the Python interpreter in the user-system environment path.

**Comment by NeoTheFox** [comment]
I see, thank you for clarification.

**Comment by Anton-Latukha** [comment]
To be proper - NixOS has the same situation with runtime deps that all other distros, and me mentioning that Nix does not have the feature - is not fair request.

Because it is for example not understandable, if qBittorrent would suddenly install Python into env - it would clash with the Python installed by the user, and it may happen that user deliberately chosen to use particular Python systemwide. Or some people never use the search engines and never need Python supplied, moreover to PATH.

Of course, we can hack everything and hack upstream projects to use hard links instead of PATH resolution - but that would resolve into huge Nixpkgs social status problems in the Open Source community, and so here Nixpkgs need to comply with upstreams to move forward with them.

**Comment by wizeman** [comment]
I would like to comment on this issue because, IMHO, I think there were some comments which reflect a lack of familiarity with the way we usually solve this problem in nixpkgs and quite a few misconceptions of how Nix actually works, so I don't think there is a fair represention of how this can be fixed cleanly in Nix/NixOS.

One of the traditional ways to solve this problem in Nix/nixpkgs/NixOS is to wrap the `qbittorrent` binary so that its `PATH` points to the exact path of a python interpreter's bin directory, regardless of whether the user already has some version of python on its `PATH` or not. In fact, this is what was done in PR #59293, which was closed without explanation.

When we wrap the qbittorrent binary to include python, different python interpreters would not in fact clash with each other, regardless of what the user has already installed or what he/she will install in the future... the entire point of building packages in Nix (or at least one of them) is that different versions of packages can all be "installed" (i.e. referenced) on the same system without clashing with each other, in fact, that's what the lengthy hashes in `/nix/store/` are for!

Note that wrapping the qbittorrent binary does not mean that installing qbittorrent would also install python into the user's environment. In fact, python would only be referenced in qbittorrent's wrapping shell script, so the user's PATH would not be changed and therefore python would not be installed into the environment. 

Also, there would be no need to hack the upstream project to use hard links instead of PATH resolution, because wrapping the binary creates a "private version" of PATH just for qbittorrent, which includes the python version specified in the Nix derivation as a higher-priority search directory and therefore overrides the version of python that the user may or may not have installed! This means that qbittorrent would not in fact end up searching the user's PATH, because it would always find python in its own PATH.

Wrapping the qbittorrent binary also means that Nix would not remove python from the qbittorrent closure, because qbittorrent's wrapping shell script would in fact reference python. So there would be no need to do anything special to "force" Nix to include python in its closure, apart from the adding the wrapping itself, which contains the reference to python because it adds it to qbittorrent's "private version" of PATH.

Note that this is something that we do very often in NixOS and nixpkgs, i.e., it's one of the standard ways of solving this issue in Nix, so no matter how "hacky" it appears to be, we have many years of experience telling us that it works very well.

In fact, forcing the user to install python in their own environment just so that qbittorrent's tracker search works is actually what can cause clashes! These clashes can happen because qbittorrent might require one python version to be in the environment, but the user may want to install and use a different python version in their environment! If the environment has 2 different versions of python, only one can actually be used, the other gets ignored, due to how PATH resolution works (unless one version is python 2.x and the other is python 3.x, because they have different binary names, but this is an exception rather than the rule).
As you mentioned, "installing" python on the environment is what is usually done on other Linux distributions to solve these runtime dependencies, but it clearly does not work well when different versions of a package need to be installed, which is why it's not a good idea to force users to install packages on their own environment in NixOS. Nix was actually designed to solve this exact problem in a cleaner way.

So wrapping the binary in a script actually solves these clashes because we're doing pure/functional package management (the Nix approach), rather than relying on the user's imperative-like `PATH` environment variable (the other distros' approach).

Now the closure size issue can be a concern, sure, but please consider that qbittorrent already has a 589 MB closure size on my system, especially due to (I imagine) QT and all its dependencies (which are needed even when compiling without GUI support), and python 3 would only add another 89 MB at most (AFAICS, python's other dependencies are shared with qbittorrent).
However, please consider that in almost all real-world NixOS installations, python 3 is already a runtime dependency of many other packages, so adding python 3 to qbittorrent, in almost all cases, would not actually increase the download size or the installed size of qbittorrent, because python 3 would already be installed. In fact, even the minimal NixOS installation ISO has python 3 in its closure!

Due to all of the above arguments, I strongly think this issue should be given another opportunity, because I think it would be solved much more cleanly using the wrapping approach than forcing the user to install python on their environment, especially due to mentioned "clashes" problem. Therefore, I am reopening this issue, but note that I am not strongly married to the wrapping approach - I also think that patching qbittorrent to point to an exact version of python is also a valid way of solving this issue, although perhaps not as clean?

I would like to finish by thanking qbittorrent's maintainers for all the work they have done. I've been using it a lot lately and it works very well, thank you!

**Comment by Anton-Latukha** [comment]
Well, I felt some straw making there, but never escape that.

To be fair, in the mentioned PR person initially was doing `sed` patching of the upstream code. So at least I directed the person to the proper solution.
I arrived into Nixpkgs when such thing was not around, or not widespread, as I always knew - runtime deps were not covered, it was an answer back in a day, and never known about usage of path expansion in the derivation, nor during reading and contributing code ever saw/draw my attention to `--prefix PATH ${makeBinPath [ python ]}";`, nor it ever was explained to me. Thou I probably not very active in Nixpkgs in last 2 years.
Nix system is vast, and was too frequently changing when I learned it, so I in some parts gave up until it settles more, after that I am in the lazy mode, and honestly, due to vastness of the practices - there should be no preliminary expectations, at least until information forms into parsable/cross-referecing Wiki. Even years ago people pointed one-another to the docs all the time, nothing changed from that time, except docs got bigger and more nuanced and there are more people, and the problem of punning everything in the sequential manual stayed.

In other words - I am thankful that my old knowledge was confronted and the current practice was explained, thou simply pointing to the solution in the right place would've been at least more effective.

So I am interested that this solution exists, fair enough, path expansion seems like a solution.

**Comment by Anton-Latukha** [comment]
https://github.com/NixOS/nixpkgs/pull/106735 addresses the question.

**Comment by Anton-Latukha** [comment]
Done.
